Core/include/Acts/Utilities/Any.hpp

```diff
@@ -0,0 +1,495 @@
+        std::cout << "- " << idx.name() << ": " << addr << std::endl;
+      }
+      std::abort();
+    }
+  }
+};
+static _AnyAllocationReporter s_reporter;
+#else
+#define _ACTS_ANY_TRACK_ALLOCATION(T, heap)
+#endif
+
+class AnyBaseAll {};
+
+/// Small opaque cache type which uses small buffer optimization
+template <size_t SIZE>
+class AnyBase : public AnyBaseAll {
+  static_assert(sizeof(void*) <= SIZE, "Size is too small for a pointer");
+
+ public:
+  template <typename T, typename... Args>
+  explicit AnyBase(std::in_place_type_t<T> /*unused*/, Args&&... args) {
+    using U = std::decay_t<T>;
+    static_assert(
+        std::is_move_assignable_v<U> && std::is_move_constructible_v<U>,
+        "Type needs to be move assignable and move constructible");
+    static_assert(
+        std::is_copy_assignable_v<U> && std::is_copy_constructible_v<U>,
+        "Type needs to be copy assignable and copy constructible");
+
+    m_handler = makeHandler<U>();
+    if constexpr (sizeof(U) <= SIZE) {
+      // construct into local buffer
+      /*U* ptr =*/new (m_data.data()) U(std::forward<Args>(args)...);
+      _ACTS_ANY_VERBOSE(
+          "Construct local (this=" << this << ") at: " << (void*)m_data.data());
+    } else {
+      // too large, heap allocate
+      U* heap = new U(std::forward<Args>(args)...);
+      _ACTS_ANY_TRACK_ALLOCATION(T, heap);
+      setDataPtr(heap);
+    }
+  }
+
+  AnyBase() {
+    m_data.fill(std::byte{0});
+#if defined(_ACTS_ANY_ENABLE_VERBOSE)
+    _ACTS_ANY_VERBOSE("Default construct this=" << this);
+#endif
+  };
+
+  template <typename T, typename = std::enable_if_t<
+                            !std::is_same_v<std::decay_t<T>, AnyBase<SIZE>>>>
+  explicit AnyBase(T&& value)
+      : AnyBase{std::in_place_type<T>, std::forward<T>(value)} {}
+
+  template <typename T>
+  T& as() {
+    static_assert(std::is_same_v<T, std::decay_t<T>>,
+                  "Please pass the raw type, no const or ref");
+    if (makeHandler<T>() != m_handler) {
+      throw std::bad_any_cast{};
+    }
+
+    _ACTS_ANY_VERBOSE("Get as " << (m_handler->heap ? "heap" : "local"));
+
+    return *reinterpret_cast<T*>(dataPtr());
+  }
+
+  template <typename T>
+  const T& as() const {
+    static_assert(std::is_same_v<T, std::decay_t<T>>,
+                  "Please pass the raw type, no const or ref");
+    if (makeHandler<T>() != m_handler) {
+      throw std::bad_any_cast{};
+    }
+
+    _ACTS_ANY_VERBOSE("Get as " << (m_handler->heap ? "heap" : "local"));
+
+    return *reinterpret_cast<const T*>(dataPtr());
+  }
+
+  ~AnyBase() { destroy(); }
+
+  AnyBase(const AnyBase& other) {
+    if (m_handler == nullptr && other.m_handler == nullptr) {
+      // both are empty, noop
+      return;
+    }
+
+    _ACTS_ANY_VERBOSE(
+        "Copy construct (this=" << this << ") at: " << (void*)m_data.data());
+
+    m_handler = other.m_handler;
+    assert(m_handler != nullptr && "Handler is null");
```

Comment:
Probably not needed anymore with above check? (same consideration applies to most other asserts in public interface.

Reply:
Probably not, I like to do too many asserts rather than too few.